api/src/main/java/org/openmrs/module/htmlformentryui/element/ObsFromFragmentElement.java

```diff
@@ -0,0 +1,361 @@
+		try {
+			fragmentParams = UiIncludeTagHandler.paramsToMap(fragmentUrlBuilder.toString());
+							
+		} catch (URISyntaxException e) {
+			throw new FragmentException("Invalid fragment URI: " + fragmentUrlBuilder.toString() , e);
+		}
+	}
+	
+	@Override
+	public void handleSubmission(FormEntrySession session, HttpServletRequest request) {
+		if (session.getContext().getMode().equals(FormEntryContext.Mode.VIEW)) {
+			return;
+		}
+		String formFieldName = getFormFieldName();
+		if (formFieldName != null) {
+			String valueText = request.getParameter(formFieldName);
+			if (StringUtils.isBlank(valueText)) {
+				return;
+			}
+			if (session.getContext().getMode().equals(FormEntryContext.Mode.EDIT)) {
+				List<Obs> existingObservations = session.getContext().removeExistingObs(concept);
+				if (CollectionUtils.isNotEmpty(existingObservations)) {
+					for (Obs obs : existingObservations) {
+						if (obs.getComment().equals(formFieldName)) {
+							session.getSubmissionActions().modifyObs(existingObservations.get(0), concept, convertToType(valueText), 
+									new Date(), null, formFieldName);
+							return;
+						}
+					}
+				}
+			}
+			session.getSubmissionActions().createObs(concept, convertToType(valueText), new Date(), null,
+					// Set the formFieldName of the widget as the Obs comment 
+					// this helps to map an Obs to it's fragment widget for cases were the same concept
+					// was used in more than one <obsFromFragment/> tag on the same form 
+					formFieldName);
+		}
+	}
+
+	@Override
+	public Collection<FormSubmissionError> validateSubmission(FormEntryContext context, HttpServletRequest request) {
+		return Collections.emptyList();
+	}
+
+	@Override
+	public String generateHtml(FormEntryContext context) {
+		String wrapperDivId = getFormFieldName() + "-wrapper";
+		initializeFragment(context);
+		StringBuilder sb = new StringBuilder();
+		try {
+			sb.append("<div id=\"" + wrapperDivId + "\">");
+			sb.append(uiUtils.includeFragment(viewProvider, fragment, fragmentParams));
+			sb.append("</div>");
+			if (FormEntryContext.Mode.VIEW == context.getMode()) {
+				// disable input widgets while in view mode
+				sb.append("<script>");
+				sb.append("jq(\"#" + wrapperDivId + " :input\").attr(\"disabled\", \"disabled\");\n");
+				if (fragment.endsWith("datetimepicker")) {
+					// if datetimepicker fragment, disable the calendar icon
+					sb.append("jq('#" + wrapperDivId + " .add-on').eq(0).off('click');");
+				}
+				sb.append("</script>");	
+			}
+			return sb.toString();
+		} catch (NullPointerException e) {
+			// if we are validating/submitting the form, then this method is being called from a fragment action method
+            // and the UiUtils we have access to doesn't have a FragmentIncluder. That's okay, because we don't actually
+            // need to generate the HTML, so we can pass on this exception.
+            // (This is hacky, but I don't see a better way to do it.)
```

Comment:
Probably not possible, but it would great if at least there was a finer-grained exception we could trap.

Reply:
Sure!